drivers/counter/timer_esp32.c

```diff
@@ -0,0 +1,290 @@
+
+static struct counter_obj_t *p_timer_obj[TIMER_GROUP_MAX][TIMER_MAX] = { 0 };
+
+static inline void counter_esp32_sem_take(const struct device *dev)
+{
+	k_sem_take(&DEV_DATA(dev)->sem, K_FOREVER);
+}
+
+static inline void counter_esp32_sem_give(const struct device *dev)
+{
+	k_sem_give(&DEV_DATA(dev)->sem);
+}
+
+static int counter_esp32_init(const struct device *dev)
+{
+	const struct counter_esp32_config *cfg = DEV_CFG(dev);
+
+	k_sem_init(&DEV_DATA(dev)->sem, 1, 1);
+
+	if (TIMG(dev) == TIMER_GROUP_0) {
+		periph_module_enable(PERIPH_TIMG0_MODULE);
+	} else if (TIMG(dev) == TIMER_GROUP_1) {
+		periph_module_enable(PERIPH_TIMG1_MODULE);
+	} else {
+		return -ENOTSUP;
+	}
+
+	if (TIMX == NULL) {
+		TIMX = (struct counter_obj_t *)k_calloc(1, sizeof(struct counter_obj_t));
+		if (TIMX == NULL) {
+			LOG_ERR("TIMER driver malloc error");
+			return -ENOMEM;
+		}
+	}
+
+	counter_esp32_sem_take(dev);
+	timer_hal_init(&TIMX->hal, TIMG(dev), TIDX(dev));
+	timer_hal_intr_disable(&TIMX->hal);
+	timer_hal_clear_intr_status(&TIMX->hal);
+	timer_hal_set_auto_reload(&TIMX->hal, cfg->config.auto_reload);
+	timer_hal_set_divider(&TIMX->hal, cfg->config.divider);
+	timer_hal_set_counter_increase(&TIMX->hal, cfg->config.counter_dir);
+	timer_hal_set_alarm_enable(&TIMX->hal, cfg->config.alarm_en);
+	if (cfg->config.intr_type == TIMER_INTR_LEVEL) {
+		timer_hal_set_level_int_enable(&TIMX->hal, true);
+	}
+	timer_hal_set_counter_value(&TIMX->hal, INITIAL_COUNT);
+	timer_hal_set_counter_enable(&TIMX->hal, cfg->config.counter_en);
+	DEV_CFG(dev)->irq_config_fn(dev);
+	counter_esp32_sem_give(dev);
+
+	return 0;
+}
+
+static int counter_esp32_start(const struct device *dev)
+{
+	counter_esp32_sem_take(dev);
+	timer_hal_set_counter_enable(&TIMX->hal, TIMER_START);
+	counter_esp32_sem_give(dev);
+
+	return 0;
+}
+
+static int counter_esp32_stop(const struct device *dev)
+{
+	counter_esp32_sem_take(dev);
+	timer_hal_set_counter_enable(&TIMX->hal, TIMER_PAUSE);
+	counter_esp32_sem_give(dev);
+
+	return 0;
+}
+
+static int counter_esp32_get_value(const struct device *dev, uint32_t *ticks)
+{
+	counter_esp32_sem_take(dev);
+	timer_hal_get_counter_value(&TIMX->hal, (uint64_t *)ticks);
+	counter_esp32_sem_give(dev);
+
+	return 0;
+}
+
+static int counter_esp32_set_alarm(const struct device *dev, uint8_t chan_id,
+				   const struct counter_alarm_cfg *alarm_cfg)
+{
+	ARG_UNUSED(chan_id);
+	uint32_t now;
+
+	counter_esp32_get_value(dev, &now);
+	counter_esp32_sem_take(dev);
+	timer_hal_set_alarm_value(&TIMX->hal, (now + alarm_cfg->ticks));
+	timer_hal_intr_enable(&TIMX->hal);
+	timer_hal_set_alarm_enable(&TIMX->hal, TIMER_ALARM_EN);
+	DEV_DATA(dev)->callback = alarm_cfg->callback;
+	DEV_DATA(dev)->user_data = (struct counter_alarm_cfg *)alarm_cfg;
+	counter_esp32_sem_give(dev);
+
+	return 0;
+}
+
+static int counter_esp32_cancel_alarm(const struct device *dev, uint8_t chan_id)
+{
+	ARG_UNUSED(chan_id);
+
+	counter_esp32_sem_take(dev);
+	timer_hal_intr_disable(&TIMX->hal);
```

Comment:
ideally you should clean `DEV_DATA(dev)->callback` which can be used to distinguish between channel being busy or not.